File: DiKErnel/Test/DiKErnel.DomainLibrary.TestUtil/ValidatorAssertHelper.cpp
// Copyright (C) Stichting Deltares 2020. All rights reserved.
//
// This file is part of DiKErnel.
//
// DiKErnel is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, version 3.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with this program. If not, see <https://www.gnu.org/licenses/lgpl-3.0.html>.
//
// All names, logos, and references to "Deltares" are registered trademarks of
// Stichting Deltares and remain full property of Stichting Deltares at all times.
// All rights reserved.

#include "ValidatorAssertHelper.h"

#include <gtest/gtest.h>

namespace DiKErnel::DomainLibrary::TestUtil
{
    void ValidatorAssertHelper::AssertBelowBound(
        const ActionWithDouble validateAction,
        const double bound)
    {
        AssertValidBound(validateAction, bound - _doubleEpsilon);
    }

    void ValidatorAssertHelper::AssertBelowBound(
        const ActionWithDouble validateAction,
        const double bound,
        const ValidationIssueType validationIssueType,
        const std::string& message)
    {
        AssertInvalidBound(validateAction, bound - _doubleEpsilon, validationIssueType, message);
    }

    void ValidatorAssertHelper::AssertEqualToBound(
        const ActionWithDouble validateAction,
        const double bound)
    {
        AssertValidBound(validateAction, bound);
    }

    void ValidatorAssertHelper::AssertEqualToBound(
        const ActionWithDouble validateAction,
        const double bound,
        const ValidationIssueType validationIssueType,
        const std::string& message)
    {
        AssertInvalidBound(validateAction, bound, validationIssueType, message);
    }

    void ValidatorAssertHelper::AssertAboveBound(
        const ActionWithDouble validateAction,
        const double bound)
    {
        AssertValidBound(validateAction, bound + _doubleEpsilon);
    }

    void ValidatorAssertHelper::AssertAboveBound(
        const ActionWithDouble validateAction,
        const double bound,
        const ValidationIssueType validationIssueType,
        const std::string& message)
    {
        AssertInvalidBound(validateAction, bound + _doubleEpsilon, validationIssueType, message);
    }

    void ValidatorAssertHelper::AssertEqualToBound(
        const ActionWithInt validateAction,
        const int bound)
    {
        AssertValidBound(validateAction, bound);
    }

    void ValidatorAssertHelper::AssertEqualToBound(
        const ActionWithInt validateAction,
        const int bound,
        const ValidationIssueType validationIssueType,
        const std::string& message)
    {
        AssertInvalidBound(validateAction, bound, validationIssueType, message);
    }

    void ValidatorAssertHelper::AssertAboveBound(
        const ActionWithInt validateAction,
        const int bound,
        const ValidationIssueType validationIssueType,
        const std::string& message)
    {
        AssertInvalidBound(validateAction, bound + 1, validationIssueType, message);
    }

    template <typename TAction, typename TBound>
    void ValidatorAssertHelper::AssertValidBound(
        const TAction validateAction,
        const TBound bound)
    {
        // Call
        const auto validationIssue = validateAction(bound);

        // Assert
        ASSERT_EQ(nullptr, validationIssue);
    }

    template <typename TAction, typename TBound>
    void ValidatorAssertHelper::AssertInvalidBound(
        const TAction validateAction,
        const TBound bound,
        const ValidationIssueType validationIssueType,
        const std::string& message)
    {
        // Call
        const auto validationIssue = validateAction(bound);

        // Assert
        ASSERT_NE(nullptr, validationIssue);
        ASSERT_EQ(validationIssueType, validationIssue->GetValidationIssueType());
        ASSERT_EQ(message, validationIssue->GetMessage());
    }
}

File: Test/DiKErnel.Util.TestUtil/ValidatorAssertHelper.cs
// Copyright (C) Stichting Deltares and State of the Netherlands 2023. All rights reserved.
//
// This file is part of DiKErnel.
//
// This program is free software; you can redistribute it and/or modify it under the terms of
// the GNU Lesser General Public License as published by the Free Software Foundation; either
// version 3 of the License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
// without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
// See the GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License along with this
// program; if not, see <https://www.gnu.org/licenses/>.
//
// All names, logos, and references to "Deltares" are registered trademarks of Stichting
// Deltares and remain full property of Stichting Deltares at all times. All rights reserved.

using System;
using DiKErnel.Util.Validation;
using NUnit.Framework;

namespace DiKErnel.Util.TestUtil
{
    public static class ValidatorAssertHelper
    {
        public const double Epsilon = 1e-6;

        public const double DoubleMin = double.MinValue;

        public const double DoubleMax = double.MaxValue;

        public const int IntegerMin = int.MinValue;

        public const int IntegerMax = int.MaxValue;

        public static void AssertValue<T>(Func<T, ValidationIssue> validationFunc, T value)
        {
            // Call
            ValidationIssue validationIssue = validationFunc(value);

            // Assert
            Assert.That(validationIssue, Is.Null);
        }

        public static void AssertValue<T>(Func<T, ValidationIssue> validationFunc, T value,
                                          ValidationIssueType validationIssueType, string message)
        {
            // Call
            ValidationIssue validationIssue = validationFunc(value);

            // Assert
            Assert.That(validationIssue, Is.Not.Null);
            Assert.That(validationIssue.ValidationIssueType, Is.EqualTo(validationIssueType));
            Assert.That(validationIssue.Message, Is.EqualTo(message));
        }
    }
}